src/nspawn/nspawn.c

```diff
@@ -3974,7 +3961,16 @@ static int outer_child(
                         return r;
         }
 
-        r = mount_move_root(directory);
+        /* Mark everything as shared so our mounts get propagated down. This is required to make new bind
+         * mounts available in systemd services inside the container that create a new mount namespace.  See
+         * https://github.com/systemd/systemd/issues/3860 Further submounts (such as /dev) done after this
+         * will inherit the shared propagation mode.
+         *
+         * IMPORTANT: Do not overmount the root directory anymore from now on to enable moving the root
+         * directory mount to root later on.
+         * https://github.com/systemd/systemd/issues/3847#issuecomment-562735251
+         */
+        r = mount_switch_root(directory, PROPAGATION_SHARED);
```

Comment:
hmm, moving this down here, will this break setup_propgate() above? i.e. we have this dir `/run/systemd/nspawn/propagate` (which probably should move to `/run/host/`, but that is still there for historical reasons), which we use as "bounce dir" for implementing "machinectl bind" and things like that. i.e. there it's MS_SLAVE to a per-container dir on the host, and whenever we mount something there from the host, it will thus appear in the container, from where we'll move it to the final location.

Previously we'd first mount the container tree shared, and then create this propagation subdir afterwards, and set it up as MS_SLAVE from the host. Swapping this around won't break this?

Reply:
There's a follow-up patch for this in the series (The third patch.).